### PR TITLE
[TRTLLM-16163][feat] Enable Nemotron 3.5 Super VL serving paths

### PR DESCRIPTION
**Pull request by eopXD**
## Description

Nemotron 3.5 Super VL could not serve with `--reasoning_parser auto`, `--tool_parser auto`,
or guided decoding. Four pieces were missing.

- **Reasoning parser.** `nemotron_h_omni` had no row in `MODEL_TYPE_TO_REASONING_PARSER`.
  Its `model_type` differs from its architecture string (`NemotronH_Omni_Reasoning_V3`) and
  the lookup is exact-match, so `auto` resolved to None.

- **Tool parser.** Same gap in `MODEL_TYPE_TO_TOOL_PARSER`. It reuses `qwen3_coder` rather
  than getting its own parser because its chat template tells the model to emit the
  Qwen3-Coder `<tool_call><function=...><parameter=...>` XML, not JSON.

- **Guided decoding.** `NemotronH_Nano_VL_V2` never defined `language_model`.
  `MultimodalModelMixin` routes four members through that property, and this wrapper
  hand-overrode two of them (`vocab_size_padded`, `infer_max_seq_len`), so nothing noticed
  the property was absent until `set_guided_decoder` was added to the mixin and landed on
  the base implementation, which raises. Guided decoding plus a one-model spec-decode mode
  (MTP) died with `NotImplementedError` at executor creation.

- **Model defaults.** `ModelLoader` gates on `hasattr(model_cls, 'get_model_defaults')`
  (`model_loader.py:712`) and applies it to the resolved *outer* class, so
  `NemotronHForCausalLM`'s `enable_block_reuse: False` never reached the user's args — block
  reuse stayed on with no Mamba state snapshot policy and no warning. The hybrid
  auto-disable in `_validate_and_adjust_mamba_snapshot_config` does not cover this: it tests
  `is_nemotron_hybrid`, which reads a flat `hybrid_override_pattern` the outer VL config
  does not carry until `post_config` reassigns `pretrained_config`. Delegating to the inner
  LM matches `modeling_qwen3_5.py:726` and `modeling_gemma4mm.py:609`.

## Test Coverage

All four run in `l0_cpu` (`unittest/_torch/modeling` at `l0_cpu.yml:46`,
`test_tool_parsers.py` at `:108`, `test_reasoning_parser.py` at `:121`), so no test-list
change is needed. All four fail on base.

| Question a reviewer will ask | Test |
|---|---|
| Does `--reasoning_parser auto` resolve now? | `test_auto_detect_nemotron_h[nemotron_h_omni]` (`tests/unittest/llmapi/test_reasoning_parser.py:951`) |
| Does `--tool_parser auto` resolve now? | `test_auto_detect_nemotron_h_omni` (`tests/unittest/llmapi/apps/test_tool_parsers.py:3133`) |
| Does guided decoding actually reach the inner LM? | `test_nemotron_nano_v2_vl_set_guided_decoder_reaches_inner_lm` (`tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py:676`) |
| Does the wrapper keep block reuse opt-in? | `test_nemotron_nano_v2_vl_defaults_keep_block_reuse_opt_in` (`tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py:650`) |

Existing coverage that guards the shared paths this PR touches:
- `TestQwen3CoderToolParser` (`tests/unittest/llmapi/apps/test_tool_parsers.py:1103`)
- `test_auto_detect_nemotron_h[nemotron_h, nemotron_h_puzzle]` (`tests/unittest/llmapi/test_reasoning_parser.py:951`)
- `test_nemotron_nano_v2_vl_model_sanity_check` (`tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py:449`)

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Added automatic `nemotron_h_omni` detection for `nemotron-v3` reasoning and `qwen3_coder` tool parsing.
- Delegated model defaults and guided-decoder setup through `NemotronH_Nano_VL_V2.language_model`.
- Removed redundant `vocab_size_padded` and `infer_max_seq_len` forwarding methods.
- The changes apply `enable_block_reuse: False` through the inner language model.

## QA Engineer Review

- Added tests for model-default delegation, block-reuse defaults, guided-decoder delegation, and parser auto-detection.
- Parser tests are listed in `tests/integration/test_lists/test-db/l0_cpu.yml`.
- The modeling test has a waiver for `test_nemotron_nano_v2_vl_video_batch_equivalence`, but the file is not listed in the test database.
- Verdict: needs follow-up.

## Per-File QA Perspective

- `tensorrt_llm/_torch/models/modeling_nemotron_nano.py`: Verify language-model delegation, guided-decoder setup, model defaults, and the removed forwarding methods.
- `tensorrt_llm/llmapi/reasoning_parser.py`: Verify exact `nemotron_h_omni` resolution to `nemotron-v3`.
- `tensorrt_llm/serve/tool_parser/tool_parser_factory.py`: Verify exact `nemotron_h_omni` resolution to `qwen3_coder`.
- `tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py`: Covers model defaults, block reuse, and guided-decoder delegation. It is not listed in `test-db/l0_cpu.yml`; one existing video-equivalence test is waived.
- `tests/unittest/llmapi/apps/test_tool_parsers.py`: Covers `test_auto_detect_nemotron_h_omni` and related Nemotron tool-parser behavior. It is listed in `test-db/l0_cpu.yml`.
- `tests/unittest/llmapi/test_reasoning_parser.py`: Covers parameterized `test_auto_detect_nemotron_h` for `nemotron_h`, `nemotron_h_puzzle`, and `nemotron_h_omni`. It is listed in `test-db/l0_cpu.yml`.

### CODE REVIEW

**Review by Wanli-Jiang — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 1]
```diff
@@ -2913,6 +2918,27 @@ def vocab_size_padded(self) -> int:
...
+        # `ModelLoader` applies `get_model_defaults()` on the resolved outer
+        # model class (this VL wrapper), not on the inner decoder, so the inner
+        # `NemotronHForCausalLM` defaults were never reaching the user's args.
+        # Delegate to keep block reuse opt-in until a Mamba state snapshot
+        # policy is configured. The hybrid auto-disable in
+        # `_validate_and_adjust_mamba_snapshot_config` cannot cover this: it
+        # tests `is_nemotron_hybrid`, which reads a flat
+        # `hybrid_override_pattern` that only appears on the inner config after
+        # `post_config` below reassigns `pretrained_config`.
+        return NemotronHForCausalLM.get_model_defaults(llm_args)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2926]: `NemotronHForCausalLM` also declares `get_preferred_kv_cache_manager_version() -> "V2"` and `get_preferred_transceiver_runtime() -> "PYTHON"`, and those are looked up on the outer class too (`llm_utils.py:637`, `:733`), so they are lost here for the same reason `get_model_defaults` was. `_Qwen3_5VLModel` delegates both alongside its defaults (`modeling_qwen3_5.py:729`) -- should this wrapper do the same, or is V1 / the fallback runtime intentional for the VL path?

File: tensorrt_llm/llmapi/reasoning_parser.py [hunk 2]
```diff
@@ -446,6 +446,7 @@ def parse(self, text: str) -> ReasoningParserResult:
     "deepseek_v4": "deepseek_v4",
     "nemotron_h": "nemotron-v3",
     "nemotron_h_puzzle": "nemotron-v3",
+    "nemotron_h_omni": "nemotron-v3",
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/llmapi/reasoning_parser.py:449]: The comment at `modeling_nemotron_nano.py:2793` says the Nano checkpoints set `model_type` equal to their architecture string, i.e. `NemotronH_Nano_Omni_Reasoning_V3` -- which has no row in this map either, and the lookup is exact-match. Does `--reasoning_parser auto` also resolve to None for that variant, or does it not emit reasoning content?

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 3]
```diff
@@ -2913,6 +2918,27 @@ def vocab_size_padded(self) -> int:
     def infer_max_seq_len(self) -> int:
         return self.llm.infer_max_seq_len()
 
+    @property
+    def language_model(self) -> torch.nn.Module:
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2915]: With `language_model` defined, the `vocab_size_padded` / `infer_max_seq_len` overrides just above are now byte-identical to `MultimodalModelMixin.vocab_size_padded` / `infer_max_seq_len`. Keeping them is exactly what hid the missing property in the first place, as the PR description notes -- suggest deleting both so the forwarding lives in one place.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 4]
```diff
@@ -2913,6 +2918,27 @@ def vocab_size_padded(self) -> int:
...
+        # guided decoding plus a one-model speculative mode (MTP) dies with
+        # `NotImplementedError` at executor creation.
+        return self.llm
+
+    @classmethod
+    def get_model_defaults(cls, llm_args: "TorchLlmArgs") -> dict:
+        # `ModelLoader` applies `get_model_defaults()` on the resolved outer
+        # model class (this VL wrapper), not on the inner decoder, so the inner
+        # `NemotronHForCausalLM` defaults were never reaching the user's args.
+        # Delegate to keep block reuse opt-in until a Mamba state snapshot
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: This comment is nearly verbatim the docstring of `test_nemotron_nano_v2_vl_defaults_keep_block_reuse_opt_in`. Consider keeping one or two lines here (hook is read off the outer class, so delegate) and leaving the longer "why the hybrid auto-disable doesn't cover this" reasoning only in the test, so the explanation lives in one place.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 3 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2915]: +1.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 4 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: +1.

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py [hunk 5]
```diff
@@ -645,6 +646,62 @@ def _build_inputs(prompts_subset, media_subset):
...
+    auto-disable in `_validate_and_adjust_mamba_snapshot_config` does not catch
+    it either: that one tests `is_nemotron_hybrid`, which reads a flat
+    `hybrid_override_pattern` the outer VL config does not carry -- the inner
+    config only becomes `pretrained_config` later, in `post_config`. Between
+    the two, block reuse stays on with no Mamba state snapshot policy and no
+    warning.
+    """
+    # `NemotronHForCausalLM.get_model_defaults` returns a constant, so the
+    # args object only has to exist.
+    llm_args = SimpleNamespace()
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py]: Nit: why not use the actual objects? I believe most fields have defaults, and as a general rule of thumb, it's usually preferable to use the actual objects than mocks.

**Review by 2ez4bz — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 1 above)
Comment (eopXD) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2926]: v2_kvcm serving currently has correctness issue. Let me follow-up this change after #18885 lands.

Comment (eopXD) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Confirmed, and the gap is real: `preference_cls` is resolved from `architectures[0]`
(`model_loader.py:705-708`) and both hooks are read off it (`llm_utils.py:636`, `:733`), so
`NemotronHForCausalLM`'s `"V2"` / `"PYTHON"` never reach the args and `auto` lands on V1 —
exactly the same mechanism as `get_model_defaults`. Both siblings this PR follows do all three
(`modeling_qwen3_5.py:729`, `:736` and `modeling_gemma4mm.py:612`, `:617`).

Not intentional, but I'd rather not flip it here. V2 still has open correctness work on this
path — #18885 (`vocab_size` never reaches `KVCacheManagerV2` on the hybrid branches, which
aborts the server on the first image once block reuse is on) is not merged yet. Switching the
manager is also class-wide: the hooks would move `NemotronH_Nano_VL_V2` and
`NemotronH_Nano_Omni_Reasoning_V3` off V1 too, so it wants its own validation run rather than
riding along with a parser-enablement PR.

Follow-up PR after #18885 lands, adding both hooks together.

File: tensorrt_llm/llmapi/reasoning_parser.py (hunk 2 above)
Comment (eopXD) [reply tensorrt_llm/llmapi/reasoning_parser.py]: Right on both counts — the lookup is exact-match (`reasoning_parser.py:542`) and
`NemotronH_Nano_Omni_Reasoning_V3` has no row, so `--reasoning_parser auto` resolves to None
for it. `NemotronH_Nano_VL_V2` is in the same position.

It's wider than reasoning: `MODEL_TYPE_TO_TOOL_PARSER` has no Nemotron row at all outside the
one this PR adds, not even `nemotron_h` / `nemotron_h_puzzle`. The shipped guidance matches —
the Nemotron 3 deployment guide has users pass both flags by hand
(`docs/source/deployment-guide/deployment-guide-for-nemotron-3-on-trtllm.md:155`).

Backlogged rather than fixed here. The rows are one line each, but each one is only correct if
that checkpoint's chat template actually emits the `<think>` shape and the Qwen3-Coder XML, and
a wrong parser fails silently — it just returns the raw text as content. That needs a serve run
per variant, which is more than this PR's scope. Filing a follow-up covering both maps for the
Nano architectures.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 3 above)
Comment (eopXD) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2915]: Agreed — with `language_model` in place both are byte-identical to the mixin
(`modeling_multimodal_mixin.py:677-683`), and `Gemma4MultimodalModelBase` already carries only
the property (`modeling_gemma4mm.py:930`). Both deleted.

Code has been updated in the latest diff.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 4 above)
Comment (eopXD) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Agreed. Trimmed to the mechanism — the hook is read off the resolved outer class, so delegate
to keep block reuse opt-in — and left the `_validate_and_adjust_mamba_snapshot_config` reasoning
in the test docstring, which is where it explains why the test exists. Did the same to the
`language_model` comment.

Code has been updated in the latest diff.

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py (hunk 5 above)
Comment (eopXD) [reply tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py]: Agreed — `TorchLlmArgs(model="/tmp/dummy_model")` now, matching
`test_qwen35_moe_model_defaults` (`tests/unittest/_torch/modeling/test_modeling_qwen3_5_vl_moe.py:194`).

Code has been updated in the latest diff.

File: tensorrt_llm/llmapi/reasoning_parser.py (hunk 2 above)
Comment (eopXD) [reply tensorrt_llm/llmapi/reasoning_parser.py:449]: Adding the model requires to run some test coverage to validate it, so it is not just pure-edit work
Will follow-up in as a separate effort.

**Review by zhaoyangwang-nvidia — APPROVED** [review]
LGTM. The two real issues are fixed in the latest diff — the duplicated `vocab_size_padded` / `infer_max_seq_len` overrides are gone, and the comment is trimmed.

Fine to defer the other two. Not flipping the KV cache manager hooks before #18885 lands is the safer call, and the missing parser rows aren't a regression from this PR.

Could you file tracking issues for both follow-ups and link them here? Mainly the KV cache one, since it's easy to lose while waiting on #18885.

**Comment by eopXD** [comment]
> Could you file tracking issues for both follow-ups and link them here? Mainly the KV cache one, since it's easy to lose while waiting on https://github.com/NVIDIA/TensorRT-LLM/pull/18885.

Thank you for the reminder.
Created JIRA ticket TRTLLM-15645 for tracking.

**Comment by eopXD** [comment]
/bot skip --comment "Skipping multi-GPU pipelines after a try. Single-GPU pipelines have passed, which covers the test case added in this MR."

File: tensorrt_llm/llmapi/reasoning_parser.py (hunk 2 above)
Comment (eopXD) [reply tensorrt_llm/llmapi/reasoning_parser.py:449]: TRTLLM-16221 to keep track of this.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 1 above)
Comment (eopXD) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2926]: TRTLLM-16220 to keep track of this.
